New memory type & explicit encoding of assignments

ThomasHaas opened the pull request:
This PR adds a new family of types `MemoryType<N>` (`N` is the number of bits) with the following operations;
- Casts between `MemoryType<N>` and any type `T` that is storable in memory with`memorySizeInBits(T) == N` (currently all our types are storable in memory, but this need not be the case in principle). A round-trip `T -> MemoryType<N> -> T` is lossless, meaning `MemoryType<N>` must be able to precisely capture any type T of appropriate size (that is storable in memory).
- Concat/Extract. Round-trips over extract/concat `MemoryType<N> -extract-> X*MemoryType<N/X> -concat-> MemoryType<N>` must be lossless, i.e., if we split the type into chunks and then concat them again correctly, we get the original object.

### The purposes of `MemoryType<N>`:

(1) Any data-flow over memory, i.e., between stores and loads, can be modelled in a more principled manner.
Conceptually, every `store(addr, val)` stores `val to MemoryType<N>` and every load `T r := load(addr)` loads a value `MemoryType<N> loadVal` and then converts it `r = loadVal to T`. By the lossless round-trip property, if the type of the load and the store match, then we get the expected semantics.
However, if the types mismatch, a sort of reinterpret happens over the intermediary memory type: `T1 -> MemoryType<N> -> T2`. This gives a unified way to deal with type mismatches/conversions over memory.

(2) Since `MemoryType<N>` supports concat/extract with lossless round-trips, we can uniformly model `Tearing` for all types:
- A store of type T gets teared like this:`T -cast-> MemoryType<N> -extract-> X*MemoryType<N/X> -store-> ...`
- A load of Type T gets teared like this: `... -load-> X*MemoryType<N/X> -concat-> MemoryType<N> -cast-> T`

This should enable automatic support for tearing of `FloatType` ( and `PointerType` #927). That being said, the current implementation does not yet encode all possible type conversions and treats `MemoryType<N>` identical to `IntegerType<N>` in the encoding.
To support lossless round-trips, the encoding of `MemoryType<N>` will likely need to become a tuple `<metadata, bits>` to keep track of additional information such as provenance, address space, or even original type.

### Distuingishing between assignments and equality

To better control where and when type conversions should occur, I added `assignEqual` methods to `ExpressionEncoder`. 
The usual `equals` methods now always encode a strict equality (no conversion) whereas `assignEqual` takes an additional parameter to define what conversion should happen (if any).
The encoding should use `assignEqual` whenever it encodes assignments, i.e., any sort of data-flow like read-from and  register assignments.

hernanponcedeleon reviewed (commented):
@ThomasHaas when I was working on adding better support for floats, I remember we discussed about some problem of floats + tearing, but I do not remember the details. Do these changes "solve" all problems, and we should easily get tearing working with floats?

Do we also need these changes to get correct semantics for #927?

ThomasHaas commented on dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ExpressionEncoder.java:
```diff
@@ -171,16 +190,25 @@ public void setEvent(Event e) {
+            checkArgument(expression.getType() instanceof MemoryType);
+            final TypedFormula<?, ?> typedFormula = encode(expression);
+            assert typedFormula.getType() == expression.getType();
+            assert typedFormula.formula() instanceof IntegerFormula || typedFormula.formula() instanceof BitvectorFormula;
```
I initially put such assertions everywhere where `MemoryType` was encoded. However, some of those assertions failed and I removed them again. However, I think I still kept those that did not fail.

The reason is that the `MemoryType` basically always occurs somewhere if you have rf-edges between mismatching types, either due to MSA (mismatching type sizes) or implicit conversions over memory (e.g., storing a float but loading it as an int).
We currently have a few such cases even when doing integer encodings.

ThomasHaas commented on dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ExpressionEncoder.java:
```diff
@@ -552,6 +580,72 @@ public TypedFormula<?, TupleFormula> visitInsertExpression(InsertExpr insert) {
+        @Override
+        public TypedFormula<?, ?> visitMemoryConcatExpression(MemoryConcat expr) {
+            checkArgument(!expr.getOperands().isEmpty());
+            Preconditions.checkState(!context.useIntegers);
```
Oh, I didn't even notice this change. I guess I accidentally let my IDE add a static import. I would rather keep the full `Precondtions.` prefix though.

ThomasHaas commented on dartagnan/src/main/java/com/dat3m/dartagnan/expression/ExpressionFactory.java:
```diff
@@ -338,7 +350,66 @@ public ScopedPointerVariable makeScopedPointerVariable(String id, ScopedPointerT
+        } else if (targetSize == sourceSize) {
+            exprMem = makeFromMemoryCast(exprMem, targetType);
+        } else {
+            assert (targetSize > sourceSize);
```
That is kinda the point. I mean, assertions are supposed to hold :P.
I just like having the condition being explicitly stated rather than inferred from the context. But that is subjective opinion.

ThomasHaas commented:
> @ThomasHaas when I was working on adding better support for floats, I remember we discussed about some problem of floats + tearing, but I do not remember the details. Do these changes "solve" all problems, and we should easily get tearing working with floats?

The idea is to streamline the handling of all data going through memory. The conceptual idea is that when logical data goes into memory it gets "serialized" into some byte representation and when it gets out of memory it gets "deserialized" into logical data again.
This serialization/deserialization process should deal (in principle) with stuff like padding, memory layout, endianness, and tearing.
All that needs to be defined per type is how that type gets serialized/deserialized. So for example, tearing of floats should work out-of-the-box as long as long as you define how conversions between `floatX` and `memX` work.
In principle it also solves type conversions between `floatX` to `iX` in the form of reinterpret casts.
That being said, the current implementation in this PR does not even perform an actual conversion, it just reinterprets a formula of type `T` as a formula of type `memX`, which works fine as long as we convert the formula eventually back to type `T` and not to some other type `T2`, i.e., we don't do type casting through memory (which does not currently happen as we basically only ever store `IntegerType` in memory).

> Do we also need these changes to get correct semantics for #927?

They are not "needed", but would make #927 a lot easier to implement because it streamlines tearing of pointers.

ThomasHaas commented:
> Btw. I'm not yet sure how to best handle types whose logical size is smaller than their memory size. For example, `bv1` should probably get converted to `mem8` rather than `mem1`. I implemented it this way, but to do the conversion, I first do a zero extension `bv1 -> bv8` and then the cast to `mem8`. Maybe it is better to first convert `bv1->mem1` and then do a zero extension `mem1->mem8`. Or maybe it is even better to just stay with `mem1` and not extend it at all? I don't know... I think both approaches are better than the current one, because not all types will inherently support the concept of zero extension.

I changed the approach now: Instead of having a loss-free cast between `bvX <-> memX`, we always cast to/from the next larger memory type whose size is a multiple of 8. So, e.g., `bv1 <-> mem8` and `bv13 <-> mem16`, reflecting the fact that each type's byte representation is a multiple of 8.

ThomasHaas commented:
The branch works better than I expected.
However, there is still one problem with `Mem2Reg` which currently inserts regular type-casts instead of bitcasts.
Apart from that, I think the branch should work.

ThomasHaas commented:
It seems the error in `float9` is due to Z3 being wrong.
So I think the PR is almost ready to be reviewed (I will do minor cleanup).
What we need to check is if the new memory cast operations make AA more imprecise or inhibit some other optimizations.

Another thing that could be done in principle is to replace the mem type by our int type for now as the memory type does not yet carry any metadata to differentiate it from ints. However, even if we did this, I would still keep the API of memory types/operations to make the intention of the operations clear.

hernanponcedeleon commented:
> So I think the PR is almost ready to be reviewed (I will do minor cleanup).

When it is really just remove the DRAFT dtatus

ThomasHaas commented on dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ExpressionEncoder.java:
```diff
@@ -552,6 +580,72 @@ public TypedFormula<?, TupleFormula> visitInsertExpression(InsertExpr insert) {
+                final IntegerFormula highBitValue = imgr.makeNumber(BigInteger.TWO.pow(expr.getHighBit() + 1));
+                final IntegerFormula lowBitValue = imgr.makeNumber(BigInteger.TWO.pow(expr.getLowBit()));
+                final IntegerFormula op = (IntegerFormula) operand;
+                final IntegerFormula extracted = expr.isExtractingHighBits() ? op : imgr.modulo(op, highBitValue);
+                enc = expr.isExtractingLowBits() ? extracted : imgr.divide(extracted, lowBitValue);
+            } else {
+                final BitvectorFormulaManager bvmgr = bitvectorFormulaManager();
+                enc = bvmgr.extract((BitvectorFormula) operand, expr.getHighBit(), expr.getLowBit());
+            }
+            return new TypedFormula<>(expr.getType(), enc);
+        }
```
I added several simplifications to `ExprSimplifier` regarding casts between memory and integer types.
The simplifications seem to be enough that without MSA and without mixed type accesses (MTA?) all memory operations get eliminated. So I added checks that `!context.useIntegers` holds whenever memory expressions are encoded.
Btw. we probably also want to check `!context.useIntegers` when dealing with floats: many operations like `isinf` or `signbit` result in float->int conversions.

hernanponcedeleon commented:
@ThomasHaas can you please fix rebase / conflicts?

hernanponcedeleon commented on benchmarks/miscellaneous/floats.c:
```diff
@@ -148,6 +144,35 @@ int main(void) {
 
+ #elif TEST_ID == 15
+     /* --- fmin vs. fmax and signed zero --- */
+     double d2 = __VERIFIER_nondet_double();
```
I would rather define `d2` at the top rather than for 15 and 16. This is consistent with e.g., `d` not being used by Test_3

ThomasHaas replied:
I changed it, but I did not bother with regenerating all float tests, meaning there is now a slight mismatch between the llvm and the C code.

hernanponcedeleon commented on benchmarks/miscellaneous/floats.c:
```diff
@@ -148,6 +144,35 @@ int main(void) {
+    /* --- fmax and signed zero corner case --- */
+    double d2 = __VERIFIER_nondet_double();
+    if (!isnan(d) && !isnan(d2)) {
+        assert (fmax(d, d2) == fmax(d, d2));
```
is this supposed to check that the SMT solver is deterministic for max? or that it can return once +0 and once -0, but then equality holds?

ThomasHaas replied:
The assertion by itself is quite pointless (although it ensures that the result is not NaN) but I put it as contrast to the coming assertions that swap the order of `d` and `d2`.
Also, the very next assertion checks sign-preservation, so those two together ensure that multiple calls to e.g. `fmax(0, -0)` consistenly return either 0 or -0.

hernanponcedeleon commented on dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ExpressionEncoder.java:
```diff
@@ -702,6 +745,109 @@ public TypedFormula<?, TupleFormula> visitInsertExpression(InsertExpr insert) {
+            final MemoryType targetType = types.getMemoryTypeFor(expr.getSourceType());
+
+            final Formula enc;
+            if (inner.getType() instanceof IntegerType iType) {
```
`inner.getType()` is used twice ... for consistency with `visitFromMemoryCastExpression` store it into a variable

hernanponcedeleon commented on dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ExpressionEncoder.java:
```diff
@@ -702,6 +745,109 @@ public TypedFormula<?, TupleFormula> visitInsertExpression(InsertExpr insert) {
+                    enc = inner.formula();
+                }
+            } else if (targetType instanceof FloatType fType) {
+                enc = floatingPointFormulaManager().fromIeeeBitvector((BitvectorFormula) inner.formula(), getFloatFormulaType(fType));
```
should we have some assert about bit width?

hernanponcedeleon commented on dartagnan/src/main/java/com/dat3m/dartagnan/encoding/ExpressionEncoder.java:
```diff
@@ -702,6 +745,109 @@ public TypedFormula<?, TupleFormula> visitInsertExpression(InsertExpr insert) {
+            final BitvectorFormulaManager bvmgr = bitvectorFormulaManager();
+            for (TypedFormula<MemoryType, ?> op : operands.subList(1, operands.size())) {
+                enc = bvmgr.concat((BitvectorFormula) op.formula(), (BitvectorFormula) enc);
+            }
```
Not sure it makes much difference, but maybe better to cast `enc` at initialization rather than at each iteration

hernanponcedeleon commented on dartagnan/src/main/java/com/dat3m/dartagnan/expression/floats/FloatBinaryOp.java:
```diff
@@ -18,8 +18,8 @@ public String getSymbol() {
             case FREM -> "%";
-            case FMAX -> "fmax ";
-            case FMIN -> "fmin ";
+            case FMAX -> "fmax";
+            case FMIN -> "fmin";
```
IIRC the spaces were on purpose ... don't we get things like `2 max1` now?

ThomasHaas replied:
No, we always added spaces before/after each binary operator. Before we got `2 max  1`.

hernanponcedeleon commented on dartagnan/src/main/java/com/dat3m/dartagnan/expression/processing/ExprSimplifier.java:
```diff
@@ -378,6 +398,82 @@ public Expression visitAggregateCmpExpression(AggregateCmpExpr expr) {
+        final List<Expression> inner = expr.getOperands().stream().map(e -> e.accept(this)).toList();
+
+        // (int x to mem) :: (int y to mem) == (int x::y) to mem
+        // TODO: only true if the int types sizes are byte-multiples
```
Does this mean this simplification is unsound otherwise? Why dont we add a check that the `bitWidth % 8 == 0` ?

ThomasHaas replied:
That's why it was a todo (maybe I should have put FIXME). It boils down to the question whether we allow the mem-type to have sizes that are not multiples of 8. If so, the optimization is sound, if not, we need to restrict it (see answer to your question below).

ThomasHaas replied:
The optimization is now sound and written so that it works with bit-granular and byte-granular memory types

hernanponcedeleon commented on dartagnan/src/main/java/com/dat3m/dartagnan/expression/type/MemoryType.java:
```diff
@@ -0,0 +1,59 @@
+
+    MemoryType(int bitWidth) {
+        Preconditions.checkArgument(bitWidth > 0, "Size for memory type must be positive: %s", bitWidth);
+        Preconditions.checkArgument((bitWidth & 7) == 0, "Size must be a multiple of 8: %s", bitWidth);
```
Why %7 and not %8?

Related to my comment about of checking this modulo ... doesn't this guarantee that memory types are always at the byte granular and those TODO can be removed?

ThomasHaas replied:
It's `& 7` not `% 7` (but I can change it to `% 8`).
The memory type is currently byte-granular (maybe we will change this at some point?).
However, int types are not byte-granular, so e.g. a `bv1` will do a round-trip like this `bv1 -> mem8 -> bv1`, which basically zero-extends the `bv1` and then truncates it again. Because of this implicit extension/truncation, some of the optimizations are not sound.
For example: `(bv1(1) :: bv1(1)) to mem != (bv1(1) to mem) :: (bv1(1) to mem)` since the former yields a `mem8` with the lowest two bits set, and the latter yields a `mem16` with the first and 9-th bit set.
It would be sound if `mem1` was a valid type though. But for now I'm inclined to assume that our memory is byte-granular.

ThomasHaas replied:
Changed to `% 8`. 
Also fixed the optimizations to be as general as possible and always sound.

hernanponcedeleon commented on dartagnan/src/main/java/com/dat3m/dartagnan/expression/ExpressionFactory.java:
```diff
@@ -209,24 +208,24 @@ public Expression makeIntegerCast(Expression operand, IntegerType targetType, bo
-    public FloatLiteral makeZero(FloatType type) {
-        return makeValue(BigDecimal.ZERO, type);
+    public FloatLiteral makePlusZero(FloatType type) {
+        return makeValue(BigDecimal.ZERO, false, type);
```
This confused me a lot ... see comment about in `FloatLiteral` class

hernanponcedeleon commented on dartagnan/src/main/java/com/dat3m/dartagnan/expression/floats/FloatLiteral.java:
```diff
@@ -74,9 +77,9 @@ public String toString() {
             return "NaN";
         }
-        return String.format("%s(%s)", getType(), value);
+        return String.format("%s(%s%s)", getType(), sign ? "-" : "", absValue);
```
I found this rather confusing. My first impression was that "true = +" and I was confused when I saw `makePlusX` in the factory get `false` as parameter.

Rather than sign it might be more clear to call it `isNegative`

ThomasHaas replied:
Technically, the cleanest variant is to use an `enum Sign { POSITIVE, NEGATIVE; }` I think. But I can also change the boolean to a `isNegative`

ThomasHaas replied:
Changed to `isNegative`

hernanponcedeleon commented:
After the changes of this PR, can we fix [this](https://github.com/hernanponcedeleon/Dat3M/blob/30640123b88d4fc5ca01403f936d45c54d83a9c1/dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java#L1803)?

ThomasHaas commented:
> After the changes of this PR, can we fix [this](https://github.com/hernanponcedeleon/Dat3M/blob/30640123b88d4fc5ca01403f936d45c54d83a9c1/dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java#L1803)?

I think the issue is unrelated to this PR but is instead related to verification with `--mixedSize=false`. With MSA detection, the check can just be removed. Without MSA, the memset is only correctly modelled for `0` and `-1` and there is no easy way around it.

hernanponcedeleon commented:
The intrinsics class knows if MSA is enabled, so we can at least remove the exception in that case

ThomasHaas commented:
I would do it in a separate PR though as there are plenty things that can be improved, for example, generating proper loops for memset/memcpy so that we support dynamic size arguments.
